### PR TITLE
Zinc compiler enhacements

### PR DESCRIPTION
**Pull request by fkorotkov**
Take number 3 after #666 and #671 on fixing timeout while locking Zinc cache.

# Background

See #666 and #671  for details.

TLDR: We have a project with 450+ project where ~100 of them in Scala. On a cold compilation where no caches are presented and we are compiling with max parallelism we see TimeoutExceptions while waiting for a zinc lock.

# Solution

I've refactored scala compilation to use direct in-memory java compiler and use separate caches for different Zinc versions. 

But most importantly I've refactored how the lock is used. Before it was acquired even on checking 
local caches. Also compilation of sbt compiler interface was wrapped in the lock which led to the timeout. Now this compilation will be take place in a temp folder and a lock will be acquired only on coping the compiler interface to the cache folder.

# Testing

I've ran v3.1.0 with this fix on our repo multiple times successfully. In our case it also dropped the overall compilation time by ~10%.

to: @ghale
cc: @eriwen @bmuschko

### CODE REVIEW

**Comment by fkorotkov** [comment]
Could someone PTAL? /cc @ghale @eriwen @bmuschko

**Comment by eriwen** [comment]
Hi @fkorotkov, I had an initial look through this. I have pushed a few minor changes to another branch 32dd734d0de3bc1cc7c770c6d34751dcd6bd62e0

I and have a couple things to follow up on:
- Understand the implications of using Zinc's compiler cache vs. Gradle's cache
- Understand the implications of using in-memory Java compiler instead of forking

Meanwhile [ScalaCompileParallelIntegrationTest](https://github.com/gradle/gradle/blob/master/subprojects/language-scala/src/integTest/groovy/org/gradle/language/scala/ScalaCompileParallelIntegrationTest.groovy) is now failing (because the interface JAR is no longer where we expect). You can run it with `./gradlew :languageScala:integTest --tests "*ScalaCompileParallelIntegrationTest"` - would you mind taking a look at it?

**Comment by fkorotkov** [comment]
@eriwen sure thing! Will fix probably tomorrow 👍

**Comment by eriwen** [comment]
@fkorotkov I've got the tests fixed and made a couple minor tweaks. Doing one final pass with my mates and we'll probably get this merged very soon :)

**Comment by fkorotkov** [comment]
@eriwen awesome news! thanks a lot!

**Comment by eriwen** [comment]
This was merged with bc8fe2fb3b300834040e46dd52ee77067fbac534 with a bit of cleanup in 92cd7a6f0ef94b74e96d587121baf48428457968.

Huge thanks @fkorotkov!

**Comment by eriwen** [comment]
@fkorotkov I'm afraid I had to revert back to forking a JVM for Zinc interface compilation (https://github.com/gradle/gradle/commit/799aa60f1474065a3f7fb2f39bc875bedbb127e1). This caused JAVA_HOME to not be detected correctly. (You can reproduce this by running `./gradlew :scala:check --tests "*JreJavaHomeScalaIntegrationTest"

Perhaps this is due to the way Zinc looks for `javac`. Would you have any ideas?

**Comment by fkorotkov** [comment]
@eriwen The in-memory one should use the same as the working process. I'm PTO this week and will take a look next week.

**Comment by fkorotkov** [comment]
@eriwen I've reverted it back to direct one and `./gradlew :scala:integTest --tests "*JreJavaHomeScalaIntegrationTest"` is working for me on Mac. What kind of failure do you see? Is it the windows only test case?

**Comment by fkorotkov** [comment]
@eriwen any updates on how can I reproduce locally?
